org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -611,7 +616,20 @@ public void disconnect(URI uri) {
 			documentListener.documentClosed();
 		}
 		if (this.connectedDocuments.isEmpty()) {
-			stop();
+			if (timer != null) {
+				timer.cancel();
+			}
+			if (this.serverDefinition.lastDocumentDisconnectedTimeout == 0) {
+				stop();
+			}
```

Comment:
I would move the code below this "if" statement (creation of the new timer etc.) into an "else" block, that makes the purpose clearer and it would make sense to be even without the test case need. If there is no timeout being set, we avoid the waste to create a new timer with a zero timeout.